# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib.pyplot import figure
import numpy as np
%matplotlib inline

# setting max columns
pd.set_option('display.max_columns', None)

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
orig_df = pd.read_csv('./data/step3_output.csv', index_col=0)

In [3]:
df = orig_df.copy()
df.head()

Name       state  vertical_drop  base_elev  \
2                Hilltop Ski Area      Alaska            294       1796   
4             Sunrise Park Resort     Arizona           1800       9200   
5   Yosemite Ski & Snowboard Area  California            600       7200   
10               Donner Ski Ranch  California            750       7031   
12                  June Mountain  California           2590       7545   

    trams  fastEight  fastSixes  fastQuads  quad  triple  double  surface  \
2       0        0.0          0          0     0       1       0        2   
4       0        0.0          0          1     2       3       1        0   
5       0        0.0          0          0     0       1       3        1   
10      0        0.0          0          0     0       1       5        2   
12      0        0.0          0          2     0       0       4        1   

    total_chairs  Runs  TerrainParks  LongestRun_mi  SkiableTerrain_ac  \
2              3  13.0           1.0            1.0               30.0   
4              7  65.0           2.0            1.2              800.0   
5              5  10.0           2.0            0.4               88.0   
10             8  52.0           2.0            1.5              505.0   
12             7  41.0           2.0            2.0             1500.0   

    Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  \
2        30.000000        150.000000       36.0             69.0   
4        80.000000        115.000000       49.0            250.0   
5       174.873239        110.000000       84.0            300.0   
10       60.000000        163.000000       82.0            400.0   
12      174.873239        115.103943       58.0            250.0   

    AdultWeekday  AdultWeekend  NightSkiing_ac  clusters  
2      30.000000      34.00000       30.000000         0  
4      74.000000      78.00000       80.000000         2  
5      47.000000      47.00000      100.395722         1  
10     75.000000      75.00000      100.395722         1  
12     57.916957      64.16681      100.395722         2

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [4]:
#Making dummy variables for the State column and dropping the original State column from the Dataframe.

df = pd.concat([df.drop(['state'], axis=1), pd.get_dummies(df.state)], axis=1)

In [5]:
df.head()

Name  vertical_drop  base_elev  trams  fastEight  \
2                Hilltop Ski Area            294       1796      0        0.0   
4             Sunrise Park Resort           1800       9200      0        0.0   
5   Yosemite Ski & Snowboard Area            600       7200      0        0.0   
10               Donner Ski Ranch            750       7031      0        0.0   
12                  June Mountain           2590       7545      0        0.0   

    fastSixes  fastQuads  quad  triple  double  surface  total_chairs  Runs  \
2           0          0     0       1       0        2             3  13.0   
4           0          1     2       3       1        0             7  65.0   
5           0          0     0       1       3        1             5  10.0   
10          0          0     0       1       5        2             8  52.0   
12          0          2     0       0       4        1             7  41.0   

    TerrainParks  LongestRun_mi  SkiableTerrain_ac  Snow Making_ac  \
2            1.0            1.0               30.0       30.000000   
4            2.0            1.2              800.0       80.000000   
5            2.0            0.4               88.0      174.873239   
10           2.0            1.5              505.0       60.000000   
12           2.0            2.0             1500.0      174.873239   

    daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  \
2         150.000000       36.0             69.0     30.000000      34.00000   
4         115.000000       49.0            250.0     74.000000      78.00000   
5         110.000000       84.0            300.0     47.000000      47.00000   
10        163.000000       82.0            400.0     75.000000      75.00000   
12        115.103943       58.0            250.0     57.916957      64.16681   

    NightSkiing_ac  clusters  Alaska  Arizona  California  Colorado  \
2        30.000000         0       1        0           0         0   
4        80.000000         2       0        1           0         0   
5       100.395722         1       0        0           1         0   
10      100.395722         1       0        0           1         0   
12      100.395722         2       0        0           1         0   

    Connecticut  Idaho  Illinois  Indiana  Iowa  Maine  Maryland  \
2             0      0         0        0     0      0         0   
4             0      0         0        0     0      0         0   
5             0      0         0        0     0      0         0   
10            0      0         0        0     0      0         0   
12            0      0         0        0     0      0         0   

    Massachusetts  Michigan  Minnesota  Missouri  Montana  Nevada  \
2               0         0          0         0        0       0   
4               0         0          0         0        0       0   
5               0         0          0         0        0       0   
10              0         0          0         0        0       0   
12              0         0          0         0        0       0   

    New Hampshire  New Jersey  New Mexico  New York  North Carolina  Ohio  \
2               0           0           0         0               0     0   
4               0           0           0         0               0     0   
5               0           0           0         0               0     0   
10              0           0           0         0               0     0   
12              0           0           0         0               0     0   

    Oregon  Pennsylvania  Rhode Island  South Dakota  Tennessee  Utah  \
2        0             0             0             0          0     0   
4        0             0             0             0          0     0   
5        0             0             0             0          0     0   
10       0             0             0             0          0     0   
12       0             0             0             0          0     0   

    Vermont 

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [6]:
# first we import the preprocessing package from the sklearn library
from sklearn.preprocessing import StandardScaler

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [7]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [17]:
#all first model set
from sklearn.linear_model import LinearRegression
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [9]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test

y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [10]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

from sklearn.metrics import explained_variance_score, mean_absolute_error

print("Explained Variance Score: ", explained_variance_score(y_test, y_pred))
print("RMSE: ", mean_absolute_error(y_test, y_pred) ** 0.5)

Explained Variance Score:  0.7972353350975028
RMSE:  2.483652654092646


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [11]:
print('Intercept value from Linear Model: ', lm.intercept_)

Intercept value from Linear Model:  18.142536478779384


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [12]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)


Coefficient
Tennessee          2.091352e+01
North Carolina     1.590843e+01
Massachusetts      1.447800e+01
Wyoming            1.243797e+01
South Dakota       1.173022e+01
Idaho              1.031926e+01
West Virginia      9.987541e+00
New Jersey         9.892482e+00
New Mexico         9.594284e+00
Arizona            9.273211e+00
Montana            8.847519e+00
Virginia           8.648856e+00
New Hampshire      8.568879e+00
Maine              8.263444e+00
Vermont            7.700161e+00
Colorado           7.340467e+00
Utah               7.197557e+00
Oregon             5.999006e+00
Missouri           5.786261e+00
Indiana            4.807222e+00
Nevada             4.766109e+00
California         4.560265e+00
Ohio               4.384494e+00
Pennsylvania       3.668068e+00
Washington         2.077023e+00
Connecticut        1.959014e+00
LongestRun_mi      1.460972e+00
Minnesota          1.237467e+00
surface            1.172606e+00
Alaska             1.116268e+00
total_chairs       1.042251e+00
Wisconsin          9.839298e-01
Iowa               9.597688e-01
Michigan           7.978445e-01
quad               7.185540e-01
clusters           7.184633e-01
AdultWeekday       6.313287e-01
fastQuads          5.826554e-01
double             5.477792e-01
New York           4.196206e-01
triple             3.658680e-01
TerrainParks       3.001861e-01
Illinois           7.926026e-02
Runs               5.912133e-02
yearsOpen          5.248694e-02
daysOpenLastYear   2.339293e-02
NightSkiing_ac     2.148932e-02
Snow Making_ac     1.206678e-02
averageSnowfall    4.323908e-03
vertical_drop      1.961044e-03
SkiableTerrain_ac  8.697739e-04
base_elev          8.606129e-04
trams              3.530806e-10
fastEight          5.987966e-11
fastSixes          2.994166e-11
Maryland           5.329071e-15
Rhode Island       4.440892e-16

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [14]:
# Liner Model without the State feature

df2 = orig_df.drop(['state'], axis=1)

In [20]:
# Declaring our Explanatory Variables for our Linear Model
X = df2.drop(['Name','AdultWeekend'], axis=1)
# Declaring our Response Variable for our model
y = df2.AdultWeekend

scaler = StandardScaler().fit(X)
X_scaled=scaler.transform(X) 

# Converting our Response Variable into a flattened array.
y = y.ravel()

# Declaring our Variables for our Train-Test_Split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

# Creating our Model and fitting out data to it
lm2 = LinearRegression()
model2 = lm.fit(X_train,y_train)

# Making our predictions with our Linear Model
y_pred2 = model.predict(X_test)

In [21]:
# Model 2's Accuracy
print("Explained Variance Score: ", explained_variance_score(y_test, y_pred2))
print("RMSE: ", mean_absolute_error(y_test, y_pred2) ** 0.5)

Explained Variance Score:  0.6837796244931496
RMSE:  2.7597496675491993


In [22]:
# Here is a Dataframe of the most important features in descending order.
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)

Coefficient
clusters           2.641707e+00
quad               1.655423e+00
surface            9.912800e-01
total_chairs       7.912474e-01
AdultWeekday       6.779846e-01
fastQuads          4.756149e-01
triple             4.461334e-01
LongestRun_mi      2.388375e-01
TerrainParks       1.702720e-01
double             9.762251e-02
Runs               6.177457e-02
yearsOpen          4.233565e-02
daysOpenLastYear   1.325575e-02
NightSkiing_ac     1.298653e-02
Snow Making_ac     8.382669e-03
averageSnowfall    5.435250e-03
vertical_drop      3.501881e-03
SkiableTerrain_ac  1.407010e-03
base_elev          1.803911e-04
fastEight          3.108624e-15
trams              3.094747e-15
fastSixes          1.887379e-15

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [ ]:
# Summit Elevation was already removed from the dataframe.
# I will now run a third model, excluding Base Elevation on this iteration.

In [23]:
df3 = df2.drop(['base_elev'], axis=1)

In [25]:
# Declaring our Explanatory Variables for our Linear Model
X = df3.drop(['Name','AdultWeekend'], axis=1)
# Declaring our Response Variable for our model
y = df3.AdultWeekend

scaler = StandardScaler().fit(X)
X_scaled=scaler.transform(X) 

# Converting our Response Variable into a flattened array.
y = y.ravel()

# Declaring our Variables for our Train-Test_Split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

# Creating our Model and fitting out data to it
lm3 = LinearRegression()
model3 = lm.fit(X_train,y_train)

# Making our predictions with our Linear Model
y_pred3 = model.predict(X_test)

In [26]:
# Model 3's Accuracy
print("Explained Variance Score: ", explained_variance_score(y_test, y_pred3))
print("RMSE: ", mean_absolute_error(y_test, y_pred3) ** 0.5)

Explained Variance Score:  0.6860331852732601
RMSE:  2.7590534769041177


In [27]:
# Here is a Dataframe of the most important features in descending order.
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)

Coefficient
clusters           3.306902e+00
quad               1.629073e+00
surface            9.996842e-01
total_chairs       8.039076e-01
AdultWeekday       6.775338e-01
fastQuads          5.223510e-01
triple             4.505673e-01
LongestRun_mi      2.278114e-01
TerrainParks       1.515288e-01
double             1.027353e-01
Runs               6.238157e-02
yearsOpen          4.369411e-02
daysOpenLastYear   1.306782e-02
NightSkiing_ac     1.292062e-02
Snow Making_ac     8.928413e-03
averageSnowfall    5.922647e-03
vertical_drop      3.464084e-03
SkiableTerrain_ac  1.407940e-03
trams              5.766221e-15
fastSixes          7.771561e-16
fastEight          2.220446e-16

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.7972353350975028|2.483652654092646|-|
| Model 2. | 0.6837796244931496|2.7597496675491993 |'state'|
| Model 3. | 0.6860331852732601|2.7590534769041177|'state','base_elev'|

Model Selection:

In [ ]:
# ________Model 3___________

# I wuld recommend Model 3 given that the resort have no feasible way of changing the location or the elevation
# of the resort. Technically the resort could potentially change the Vertical Drop by making their Runs go farther
# down their mountain. They also could change any of the other features in this dataframe.